src/facade/dragonfly_connection.cc

```diff
@@ -611,8 +618,17 @@ auto Connection::ParseRedis() -> ParserStatus {
       bool is_sync_dispatch = !cc_->async_dispatch && !cc_->force_dispatch;
       if (dispatch_q_.empty() && is_sync_dispatch && consumed >= io_buf_.InputLen()) {
         // Gradually release the request pool.
+        // The request pool is shared by all the connections in the thread so we do not want
+        // to release it aggressively just because some connection is running in
+        // non-pipelined mode. So we wait at least N times,
+        // where N is the number of connections in the thread.
         if (!free_req_pool_.empty()) {
-          free_req_pool_.pop_back();
+          ++free_req_release_weight;
+          if (free_req_release_weight > stats_->num_conns) {
+            free_req_release_weight = 0;
+            stats_->pipeline_cache_capacity -= free_req_pool_.back()->StorageCapacity();
+            free_req_pool_.pop_back();
+          }
```

Comment:
I forgot, do you release them to reduce memory usage after peaks? Or for some other reason? 
Your current mechanism doesn't look very reliable for this case 🤔

Reply:
The reason is pragmatic. I want to keep the cache if it's need and free it if it's not. This heuristic is simple, indeed but it will gradually release cache if it's not being used and we share the cache among all connections so it won't overuse it.